### ISSUE TITLE
Show XML doc for enum fields in external metadata

### ISSUE THREAD

**Issue by kerams** [dotnet/fsharp#17939]
https://github.com/user-attachments/assets/ce8885db-9de2-49c1-834f-c63a73d698f8

It is displayed in tooltips, but not in the decompiled view.

**Additional context**

17.12.0 Preview 4.0

**Comment by vzarytovskii** [comment]
Yeah, we probably don't traverse or generate those when we generate decompiled view. Should be an easy fix (famous last words).

**Comment by ThobiWanKenobi** [comment]
I am interested in trying to accomplish this! Does anyone have any great tips for getting started with my first issue or where to start? I am about to read the developer guide and some of the compiler documentation.

**Comment by majocha** [comment]
Awesome! Hi, @ThobiWanKenobi, `FSharpEntity.TryGetMetadataText()` is the entry point in the service for this feature. I guess that's where the investigation starts.

As for the workflow, I'd add a small test case in `FSharp.Compiler.Service.Tests.Symbols` for this.

**Comment by ThobiWanKenobi** [comment]
Okay, I do have two questions. First of all, regarding the workflow. Lets say that I want to work on this, I am trying to set a breakpoint, and want to "activate" the line. What solution would I need to build, or what project would I need to run to be able to test this? And finally, I am experiencing a build issue, since the global.json file expects a .NET preview which is not yet publicly available from the downloads page (Specifically 10.0.100-preview.7.25322.101, but only 10.0.100-preview.6) is available for download , how do I handle that?

Thanks for the patience and helping me getting started!

**Comment by majocha** [comment]
Simple fix for the dotnet issue that works for me, assuming you're using Visual Studio:
Once you build from command line e.g. with `Build.cmd -c Debug`, you'll have a `.dotnet` folder in your repo root. Just copy its contents into `C:\Program Files\dotnet`. Visual Studio will see it then.

Working on this particular issue:

1.  If you want to see it in action in actual IDE. You need to build VisualFSharp.sln in debug mode.  Ctrl-T search for `FSharpEntity.TryGetMetadataText` and put a breakpoint there, then start debugging VisualFSharpDebug project. You'll see an experimental instance of VS, open some F# code and F12 on external symbol should activate your breakpoint.

2. Another, faster option is to write a test and debug it. You can use FSharp.Compiler.Service.sln for it. This is cross platform and should work in other IDEs.

**Comment by ThobiWanKenobi** [comment]
Thanks for your help! I will begin working on this and see what I can make up! I will text if I run into more problems

**Comment by ThobiWanKenobi** [comment]
Alright so, I have managed to get tests running, tried to write a small test, for what I was recommended. I came up with the following test, let me know if there is something I am doing wrong in here. However, this test right here does pass, which leads me to believe, that this entry point, we were discussing might now have been the correct one? My question is, when do we generate the "decompiled view"? Because this function does seem to at least recursive on the fields themselves to my understanding. The test I have written can be found below

```
    [<Fact>]
    let ``XmlDocSig of enums`` () =
        let source = """
module Mod1

///<summary>Enumerates values returned by several types.</summary>
///<remarks><para>Portions of this page are modifications based on work created and shared by the....</para></remarks>
[<Struct>]
type MyEnum = 
    | Test1 = 0
    | Test2 = 3
    | Test3 = 4
    | Test4 = 2
    | Test5 = 1

let val1 = MyEnum.Test1
"""
        let fileName, options = mkTestFileAndOptions [| |]
        let _, checkResults = parseAndCheckFile fileName source options  

        let mod1 = checkResults.PartialAssemblySignature.FindEntityByPath ["Mod1"] |> Option.get
        let myEnum = checkResults.PartialAssemblySignature.FindEntityByPath ["Mod1"; "MyEnum"] |> Option.get
        mod1.XmlDocSig |> shouldEqual "T:Mod1"

        // Check XmlDocSig for each enum case
        let expectedCases =
            [ "Test1", "P:Mod1.MyEnum.Test1"
              "Test2", "P:Mod1.MyEnum.Test2"
              "Test3", "P:Mod1.MyEnum.Test3"
              "Test4", "P:Mod1.MyEnum.Test4"
              "Test5", "P:Mod1.MyEnum.Test5" ]

        for field in myEnum.FSharpFields do
            match List.tryFind (fun (name, _) -> name = field.Name) expectedCases with
            | Some (_, expectedSig) -> field.XmlDocSig |> shouldEqual expectedSig
            | None -> () // skip compiler-generated fields like "value__"
```

**Comment by majocha** [comment]
My understanding is that the issue is about xml doc missing from metadata text of a particular enum field.
Example input would be:
```fsharp
type MyEnum =
    | First = 1
    /// <summary> Some documentation for Second </summary>
    | Second = 2
``` 

Test could look something like this:

```fsharp
module MetadataAsText =
    [<Fact>]
    let ``TryGetMetadataAsText returns metadata for enum field`` () =
        let _, checkResults = getParseAndCheckResults """

type MyEnum =
    | First = 1
    /// <summary> Some documentation for Second </summary>
    | Second = 2
"""
        let symbolUse = checkResults |> findSymbolUseByName "MyEnum"
        match symbolUse.Symbol with
        | :? FSharpEntity as symbol ->
            match symbol.TryGetMetadataText() with
            | Some metadataText ->
                // To see what was produced:
                printfn $"{metadataText}"

                // TODO: Assert result contains "Some documentation for Second"

            | None ->
                failwith "Expected metadata text, got None"
        | _ -> failwith "Expected FSharpEntity symbol"
```

You would probably see that the generated metadata is missing the doc:

```
type MyEnum =
    | First = 1
    | Second = 2
```

**Comment by ThobiWanKenobi** [comment]
> MetadataAsText

Well, I would have thought the same now that you mentioned it, but now I just become even more unsure what this Issue is actually related to. If you look in `XmlDocTests.fs` on line 1401 to line 1505, there is already a passing test, that does indeed check that fields on enums are properly documented. 

The name of this issue also mentions specifically "external metadata", could it be, that the Xml documentation does work, for "internal" metadata, but not for whatever "external" metadata means?

**Comment by majocha** [comment]
Yes, external metadata means symbols that come e.g. from referenced dlls. To give an example relevant to this particular issue, you can write
```
let kind = System.DateTimeKind.Utc
```
It's an external symbol and the enum field has some xmldoc:

<img width="1174" height="183" alt="Image" src="https://github.com/user-attachments/assets/51920d2f-785e-4d87-ae24-eab01db2f725" />

When you do F12 (go to definition) on the symbol, you get the generated external metadata text:

<img width="1350" height="489" alt="Image" src="https://github.com/user-attachments/assets/3f2ad8a2-9e44-4fce-b950-9ea103212597" />

but the xml doc for `Utc` is not there.

**Comment by ThobiWanKenobi** [comment]
Okay, so in the above mentioned case, the second screenshot shows the generated external metadata text, but we expected something like 

```
[<Struct>]
type DataTimeKind = 
   | Unspecified = 0
   /// The time represented is UTC.
   | Utc = 1
   | Local = 2
```

Of course, the other fields would probably also have some, but that is the idea? Sadly, I cannot get VisualFsharp to work, everytime I try to open any code in it, it keeps crashing, so I cannot in any way reproduce it beyond testing it, it seems.

**Comment by majocha** [comment]
> Okay, so in the above mentioned case, the second screenshot shows the generated external metadata text, but we expected something like
> 
> ```
> [<Struct>]
> type DataTimeKind = 
>    | Unspecified = 0
>    /// The time represented is UTC.
>    | Utc = 1
>    | Local = 2
> ```
> 
> Of course, the other fields would probably also have some, but that is the idea? 

Yes, exactly.



>Sadly, I cannot get VisualFsharp to work, everytime I try to open any code in it, it keeps crashing, so I cannot in any way reproduce it beyond testing it, it seems.

This is probably https://github.com/dotnet/fsharp/issues/18766, a temporary issue and the workaround is in the link.

**Comment by ThobiWanKenobi** [comment]
Thank you, it seems I did not do enough research on the problem to encounter this!

I was not aware that it was a generated view you got when you pressed F12 on symbols, I always thought it opened the actual files 😅 Now I understand the problem much better, and I can in fact see it for myself! Thank you

**Comment by baronfel** [comment]
There _is_ a mode where f12 gets you actual symbols in various tooling experiences - this is powered by SourceLink metadata embedded in the PDBs for the library you're triggering the action on. I'm not sure the level of support for that in F# in VS, but that _is_ a thing that exists in the broader F# and .NET tooling ecosystem.

**Comment by kerams** [comment]
> I'm not sure the level of support for that in F# in VS

I managed to see SourceLink code and even step into it from F# code exactly once, on February 2 2025, that fateful day. It's always been decompiled F# metadata before and since (I wonder whether there's a setting that's permanently disabled on my end).

**Comment by ThobiWanKenobi** [comment]
Thanks for enlightening me with all this new info! I appreciate it. I have at least now managed to get the basic idea, and trigger this using a breakpoint, which currently is how I am most comfortable trying to work with this! I got VisualFsharp to work, and can better begin to grasp the issue!

**Comment by ThobiWanKenobi** [comment]
I got an update! I have been able to make the decompiled view show the metadata, but the tooltip does not. I was wondering whether this is intentional, because it actually takes a different execution path, or if that is part of the issue? See the attached screenshots.

However, interestingly, if I attempt to create and run a test, similar to the one proposed by majocha, the test fails, as the xmldoc is not part of that result 🤔

<img width="1761" height="673" alt="Image" src="https://github.com/user-attachments/assets/7e1da569-0721-4041-9cac-3f2c16765491" />
<img width="1110" height="440" alt="Image" src="https://github.com/user-attachments/assets/c2761cc9-62a6-41d2-8afd-53257dfa1364" />

**Comment by majocha** [comment]
Great news! Tooltip is ok as is, not an issue. To see xml doc for a particular enum case you hover over it (for example Utc) and it shows only the relevant doc.

It seems from the first screenshot that `TryGetMetadataText` already emits the text fine. Probably you need to look at the test output / debug it.

**Comment by ThobiWanKenobi** [comment]
Also, I did some further digging. It turns out, that internally there is a difference between types that come from IL (System.DateTimeKind) is an example of this, which to my understanding should work correctly, with the changes I have made. However, in the test, we are instead dealing with a type defined "in F#" code directly, and does not come from IL, and therefore is treated rather differently.

So the question is, how they differ and how we deal with it. I tried F12 on the enum defined directly in F#, just as in the test, but it does not seem not actually "go anywhere", it seems to be stuck claiming it is navigating. So its a completely different case, apparently, than getting external metadata to show, and I am unsure about the intended behaviour of this.

TLDR; The above test, actually tests for "internal metadata" which is differently handled from "external metadata" and therefore does not pass, also, does not seem to navigate anywhere, at least when I try to get it to.

**Comment by majocha** [comment]
Yeah, we should concentrate on external symbols for this issue. Navigation getting stuck is another bug, worth opening a separate issue.

**Comment by ThobiWanKenobi** [comment]
Great! In that case, I will just review my changes and ensure every change makes sense and see if I can open a PR with my changes.

**Comment by ThobiWanKenobi** [comment]
Does anyone know if I have to do something to get my PR reviewed? Or if I just need to be patient. Also I have this ILVerify ci failing, and I am not sure why... 🤔

**Comment by majocha** [comment]
> Does anyone know if I have to do something to get my PR reviewed? Or if I just need to be patient.

Yeah, it takes time :)

> Also I have this ILVerify ci failing, and I am not sure why... 🤔

Check https://github.com/dotnet/fsharp/blob/main/DEVGUIDE.md#updating-ilverify-baselines

**Comment by ThobiWanKenobi** [comment]
Okay, is there anywhere I can read about what these baselines even are, or can you explain what it is? Because I don't understand what this baseline that I probably need to update is. What is its purpose, and why do we need it?

Also thanks for the quick replies, I will wait patiently for a review then, after I figure out this baseline stuff! And hopefully it will stop failing in the pipeline.

**Comment by Martin521** [comment]
Here is what I have gathered from comments of others over the last year.
ILVerify is a tool that verifies IL code to be correct according to the spec (standard).
Therefore, the output should really be empty (no errors).
We run it as part of regression testing.
Unfortunately, the F# compiler produces some code that ILVerify moans about. This should be fixed, but has not yet been fixed.
So, in the meantime, we just keep track of those error messages in a baseline file and compare new ILVerify output with the baseline.
To then manually judge if we have added some really serious error. (Most often, only line numbers have shifted.)
And if there are changes that all seem to be ok, replace the baselines by the new output.

**Comment by ThobiWanKenobi** [comment]
Interesting! I am wondering, how and why we can accept such errors to occur? I would have thought that a compiler, that does not generate code that lives up to spec would be seen as some medium to high priority problem. How does it still work despite not living up to spec?

**Comment by edgarfgp** [comment]
Closing this as was fixed by https://github.com/dotnet/fsharp/pull/18800
